contracts/tgrade-valset/src/contract.rs

```diff
@@ -523,6 +529,45 @@ fn list_active_validators(
     })
 }
 
+fn list_jailed_validators(
+    deps: Deps,
+    env: Env,
+    start_after: Option<String>,
+    limit: Option<u32>,
+) -> Result<ListValidatorResponse, ContractError> {
+    let cfg = CONFIG.load(deps.storage)?;
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let start_after = maybe_addr(deps.api, start_after)?;
+    let start = start_after.map(|addr| Bound::exclusive(addr.as_str()));
+
+    let validators: Vec<OperatorResponse> = operators()
+        .range(deps.storage, start, None, Order::Ascending)
+        .map(|r| {
+            let (operator, info) = r?;
+
+            let jailed_until = JAIL
+                .may_load(deps.storage, &Addr::unchecked(&operator))?
+                .filter(|expires| !(cfg.auto_unjail && expires.is_expired(&env.block)));
```

Comment:
I would reverse these nested iterators. First range over `JAIL` and then, if jailed, load the required data from `operators()`.

The reason being that at any given time, there will be far fewer jailed than operators.